From 954b3fc6bce0805bc2542b70a58dd732fa6cf7dc Mon Sep 17 00:00:00 2001
From: juwei <juwei@codeaurora.org>
Date: Fri, 30 Mar 2018 18:04:49 +0800
Subject: [PATCH 18/19] Settings: Sim status and IMEI info are wrong if only
 sim2 inserted

Framework only returns the info of subscription which active
currently, not all the subscription info. so we must match the
slot position to find the right info from the list, otherwise
settings will FC if click IMEI2.

Change-Id: I1cd08ee7208cacec7ee83eac430715e49e163ce4
CRs-Fixed: 2211867
---
 .../imei/ImeiInfoDialogController.java        | 19 ++++++++++++++-----
 .../simstatus/SimStatusDialogController.java  | 10 +++++++---
 2 files changed, 21 insertions(+), 8 deletions(-)

diff --git a/src/com/android/settings/deviceinfo/imei/ImeiInfoDialogController.java b/src/com/android/settings/deviceinfo/imei/ImeiInfoDialogController.java
index e88a39a0c4..57603a5bb4 100644
--- a/src/com/android/settings/deviceinfo/imei/ImeiInfoDialogController.java
+++ b/src/com/android/settings/deviceinfo/imei/ImeiInfoDialogController.java
@@ -89,7 +89,8 @@ public class ImeiInfoDialogController {
 
     private void updateDialogForCdmaPhone() {
         final Resources res = mDialog.getContext().getResources();
-        mDialog.setText(ID_MEID_NUMBER_VALUE, getMeid());
+        mDialog.setText(ID_MEID_NUMBER_VALUE,
+                mSubscriptionInfo != null ? getMeid() : "");
         mDialog.setText(ID_MIN_NUMBER_VALUE,
                 mSubscriptionInfo != null ? mTelephonyManager.getCdmaMin(
                         mSubscriptionInfo.getSubscriptionId()) : "");
@@ -114,9 +115,13 @@ public class ImeiInfoDialogController {
     }
 
     private void updateDialogForGsmPhone() {
-        mDialog.setText(ID_IMEI_VALUE, getTextAsDigits(mTelephonyManager.getImei(mSlotId)));
+        mDialog.setText(ID_IMEI_VALUE,
+                mSubscriptionInfo != null ?
+                getTextAsDigits(mTelephonyManager.getImei(mSlotId)) : "");
         mDialog.setText(ID_IMEI_SV_VALUE,
-                getTextAsDigits(mTelephonyManager.getDeviceSoftwareVersion(mSlotId)));
+                mSubscriptionInfo != null ?
+                getTextAsDigits(mTelephonyManager.
+                        getDeviceSoftwareVersion(mSlotId)) : "");
         // device is not CDMA, do not display CDMA features
         mDialog.removeViewFromScreen(ID_CDMA_SETTINGS);
     }
@@ -127,8 +132,12 @@ public class ImeiInfoDialogController {
         if (subscriptionInfoList == null || subscriptionInfoList.size() <= slotId) {
             return null;
         }
-
-        return subscriptionInfoList.get(slotId);
+        for (SubscriptionInfo info : subscriptionInfoList) {
+            if (slotId == info.getSimSlotIndex()) {
+                return info;
+            }
+        }
+        return null;
     }
 
     @VisibleForTesting
diff --git a/src/com/android/settings/deviceinfo/simstatus/SimStatusDialogController.java b/src/com/android/settings/deviceinfo/simstatus/SimStatusDialogController.java
index 7cbf318fef..76daa6251d 100644
--- a/src/com/android/settings/deviceinfo/simstatus/SimStatusDialogController.java
+++ b/src/com/android/settings/deviceinfo/simstatus/SimStatusDialogController.java
@@ -409,11 +409,15 @@ public class SimStatusDialogController implements LifecycleObserver, OnResume, O
     private SubscriptionInfo getPhoneSubscriptionInfo(int slotId) {
         final List<SubscriptionInfo> subscriptionInfoList = SubscriptionManager.from(
                 mContext).getActiveSubscriptionInfoList();
-        if (subscriptionInfoList != null && subscriptionInfoList.size() > slotId) {
-            return subscriptionInfoList.get(slotId);
-        } else {
+        if (subscriptionInfoList == null) {
             return null;
         }
+        for (SubscriptionInfo info : subscriptionInfoList) {
+            if (slotId == info.getSimSlotIndex()) {
+                return info;
+            }
+        }
+        return null;
     }
 
     @VisibleForTesting
-- 
2.17.1

